ArduPlane/mode.h

```diff
@@ -0,0 +1,508 @@
+#pragma once
+
+#include <stdint.h>
+
+class Mode
+{
+public:
+
+    // Auto Pilot modes
+    // ----------------
+    enum Number {
+        MANUAL        = 0,
+        CIRCLE        = 1,
+        STABILIZE     = 2,
+        TRAINING      = 3,
+        ACRO          = 4,
+        FLY_BY_WIRE_A = 5,
+        FLY_BY_WIRE_B = 6,
+        CRUISE        = 7,
+        AUTOTUNE      = 8,
+        AUTO          = 10,
+        RTL           = 11,
+        LOITER        = 12,
+        AVOID_ADSB    = 14,
+        GUIDED        = 15,
+        INITIALISING  = 16,
+        QSTABILIZE    = 17,
+        QHOVER        = 18,
+        QLOITER       = 19,
+        QLAND         = 20,
+        QRTL          = 21,
+        QAUTOTUNE     = 22,
+    };
+
+    // Constructor
+    Mode();
+
+    // enter this mode, always returns true/success
+    bool enter();
+
+    // perform any cleanups required:
+    void exit();
+
+    // returns a unique number specific to this mode
+    virtual Number mode_number() const = 0;
+
+    // returns full text name
+    virtual const char *name() const = 0;
+
+    // returns a string for this flightmode, exactly 4 bytes
+    virtual const char *name4() const = 0;
+
+    //
+    // methods that sub classes should override to affect movement of the vehicle in this mode
+    //
+
+    // convert user input to targets, implement high level control for this mode
+    virtual void update() = 0;
+
+    //
+    // attributes for mavlink system status reporting
+    //
+    // return if in non-manual mode : Auto, Guided, RTL
+    virtual bool is_autopilot_mode() const { return false; }
+
+    // true if this mode accepts guided control commands
+    virtual bool is_guided() const { return false; }
+
+    // returns true if any RC input is used
+    virtual bool has_manual_input() const;
+
+    // true if heading is controlled
+    virtual bool attitude_stabilized() const { return true; }
+
+    // true only for vtol related FlightModes
+    virtual bool is_vtol_flightmode() const { return false; }
+
+    virtual bool allows_stabilized_stick_mixing() const { return true; }
+
+protected:
+
+    // subclasses override this to perform checks before entering the mode
+    virtual bool _enter() { return true; }
+
+    // subclasses override this to perform any required cleanup when exiting the mode
+    virtual void _exit() { return; }
+
+    bool q_enter();
+    void q_update();
+
+    bool guided_enter();
+    void guided_update();
+
+    void fbwa_update();
```

Comment:
Why extract this? Why not allow the modes to interact more directly? (Or in the Autotune case that this is for, just inherit from mode `ModeFBWA` instead of `Mode`)

Reply:
This first pass is meant to be simplified and straight forward. Next onion pass will introduce the inheritance stuff better.